Polymer class revamp

chrisjonesBSU opened the pull request:
### PR Summary:
For some background, I'm currently working on a project where we need to easily initialize polydisperse polymer systems, so I spent quite a bit of time playing around with `polymer.py`.  The `Polymer` class worked great at building up polymers from monomers, but in order to automate the process it required some work "preparing" the monomers before using `Polymer()` then some steps to "finish" the polymer compounds (basically putting back any missing hydrogens). 

So, I thought it might be worth a shot to implement these "pre" and "post" polymer steps into the `Polymer` class itself while making it functional for just about any monomer compound that's initialized from either a SMILES string or a file. 

**I included a jupyter notebook that shows an example of how to go through the process and describing what is happening.** It's currently located  at `mbuild/mbuild-polymer-example.ipynb`. I wasn't sure if there was a better way to attach or include an interactive notebook, so I just included it in the PR.  If this gets merged, I'll make sure to remove the notebook and make a separate PR to the tutorials repo.

Also, I merged in the changes from #839 to this PR as I needed to use some of the functionality from it.

I'm interested in any feedback, I imagine this is more in the rough draft stages than near completion, but hopefully it's a start to making it easier to build-up polymers.

### PR Checklist 
------------
 - [ ] Includes appropriate unit test(s)
 - [ ] Appropriate docstring(s) are added/updated
 - [ ] Code is (approximately) PEP8 compliant
 - [ ] Issue(s) raised/addressed?

chrisjonesBSU commented:
The most recent commit made some changes to the `add_end_groups` and `build` functions and added some doc strings.

Previously, if you added an end group compound, it was automatically duplicated, so that the same structure was at the head and tail of the polymer. Now, there is a `duplicate` parameter in `add_end_groups` that lets you keep that behavior, or you can only add the compound once. You can call the function again to add a second compound.  There's also a couple checks to ensure that `self.end_groups` never contains more than 2 compounds.

Also, the default behavior of `build` was to use hydrogen atoms as the "end groups" if `self.end_groups` was empty. Now there is a `add_hydrogens` parameter that lets the user keep that behavior, or if set to False (and self.end_groups is empty) then nothing is added to each end of the polymer.

I think there is still some work to do with how end groups are handled. Specifically if the user only wants one end group rather than 2. Right now, I'm pretty sure that would throw an error.

jennyfothergill commented on mbuild/lib/recipes/polymer.py:
```diff
@@ -12,39 +13,61 @@
+        by calling the add_monomer and add_end_group methods.
     """
-    def __init__(self, monomers, n, sequence='A', port_labels=('up', 'down')):
+    def __init__(self):
```
Thinking about this a little more, what if you had to pass in monomers with two ports (e.g., 'up' and 'down') 
the `add_monomer` could be changed to `create_monomer` and its function could be to replace specific atoms with ports or to add ports. This would allow for the existing CH2() and CH3() classes to be used more easily.

Same for the end_groups, this would allow them to be passed in with the init.

chrisjonesBSU replied:
I'm not opposed to having `monomers` and `end_groups` as parameters of the `init` function. I think that could be useful for someone that wants to go through the process of creating their own ports instead of using `add_monomer` and `add_end_groups`, and basically going straight to calling the `build` function.  There could be a check that asserts the required ports are in place.

Technically, the `CH2` class does still work with the planned API as is:

```
from mbuild.lib.moieties import ch2
chain = Poly.Polymer()
chain.add_monomer(compound = ch2.CH2(),
                 bonding_indices = [0, 0],
                 separation=0.15,
                 orientation = [(0,1,0), (0,-1,0)],
                 replace=False)
chain.build(n=5)
```

When I run this, it basically reproduces what `alkane.py` does.

jennyfothergill replied:
There may be an issue with this: CH2() already has ports and add_monomer adds more ports on top of them, which I think is not ideal? 
```
chain.visualize(show_ports=True).show()
print(len(chain.all_ports()))  # 10
print(len(chain.available_ports()))  # 0
```

chrisjonesBSU replied:
Yeah, I think having the option to skip the `add_monomer` and `add_end_groups` functions and go straight to `build` while passing in compounds that already have ports is a good idea.

In regards to the mbuild library moieties, I think it would be nice to have the option to not include ports. Something like `ch2 = CH2(add_ports=False)` but maybe that can be a separate issue, and the changes to the `Polymer` class can be made to accommodate compounds with preexisting ports.

chrisjonesBSU commented on mbuild/lib/recipes/polymer.py:
```diff
@@ -56,20 +79,138 @@ def __init__(self, monomers, n, sequence='A', port_labels=('up', 'down')):
+        separation : float, units nm
+            The bond length desired at the monomer-monomer bonding site.
+            (separation / 2) is used to set the length of each port
+        port_labels : list of str of length 2, default=['A', 'B']
```
I agree...I was also thinking about removing the labels as a parameter all together, since it doesn't really matter what they are called (at least that I can think of). So, being able to change them wouldn't make the workflow any different or better. It might only cause issues/confusion. For example, there is a check to make sure the port labels of each monomer are the same.

rsdefever commented on mbuild/lib/recipes/polymer.py:
```diff
@@ -100,28 +109,29 @@ def build(self, n, sequence='A'):
+            head_port.update_separation(0.0547) # 1/2 H-C bond len
             tail_port.update_separation(0.0547)
-        else: # Use compounds in self.end_groups
+        elif self.end_groups: # Use compounds in self.end_groups
```
Does this need to be an `elif` or can it go in the `if` loop below?

chrisjonesBSU replied:
You're right, nice catch. Thanks!  I cleaned up this section a little bit in the latest commit.

jennyfothergill commented:
@justinGilmer here is an MWE of what I was talking about with the ports:
```
from mbuild.lib.recipes import Alkane

#propyl = Alkane(3, cap_end=False)           # cap_end has the same result
#propyl.visualize(show_ports=True).show()
propyl = Alkane(3, cap_front=False)
propyl.visualize(show_ports=True).show()

print([p.used for p in propyl.all_ports()])  # [False, False, False]
print(len(propyl.available_ports()))         # 1
```
To recreate the behavior of Alkane from before, we expect one port at the uncapped end, but there are three ports in each Alkane instance.

chrisjonesBSU commented:
I had a ton of help from @jennyfothergill to make the `Polymer` class much more flexible, and usable with compounds that already have ports.

- Add different end groups
- Leave the end groups off to leave open ports on each end of the polymer. 
- Pass in compounds when creating a `Polymer` instance, and go straight to the `build` function and not use the `add_monomer` and `add_end_group` functions
- You can combine both approaches, passing in compounds to the `Polymer` instance, but still add more monomers and/or end groups using the built in functions. 

The new options and functionality are shown in the example notebook.

bc118 commented:
I think it would be good to add more information on the Polymer.add_end_groups.index docSting.  You may even consider doing a hack draw of an example in the docString (See below -> Element_AtomNumber )

         H_1        H_4 
            \           /
 H_2 - C_1 - C_2 - H_5 
            /          \
          H_3      H_6 
  
The graphic you send to me and presented in the dev meeting would also be good to include in the examples or Jupiter notebook, as it is a great visual.

bc118 commented:
For the Polymer.build  docSting, I would just add a example in text there.  
- if there are 2 monomers provided sequence = "AB" or "BAAB" etc.  
 - if there are 3 monomers provided sequence = "ABC" or "BCAB" etc.  
- if only capital or lowercase are accepted.

bc118 commented:
In general, at least for the new functions you may want to add error checking, with more specific errors, which is more targeted for new users.  

For the example below: If you use an incorrect value for the indices (float instead of int), you get a generic error "IndexError: list index out of range".   We may want to check the inputs to see if they are the correct type in the general context of new functions.  

chain.add_monomer(compound=comp,
                  indices=[13.0, -3],
                  separation=.15,
                  replace=True)
                  
For separations check if int or float.  For replace check if bool....

bc118 commented:
For the below example, I am getting different molecule build coordinates for every time I run it.  I'm not sure if we need/expect the identical build every time?  Maybe I missed it, but is there a random seed default or settable?  Maybe I'm not running it properly?

**Example (1,3-Dipropylcyclohexane):**

comp = mb.load('CCCC1CCCC(C1)CCC', smiles=True) # mBuild compound of the monomer unit
chain = Polymer()

chain.add_monomer(compound=comp,
                  indices=[13, -3],
                  separation=.15,
                  replace=True)

chain.build(n=10, sequence='A')
chain.visualize(show_ports=True).show()

chrisjonesBSU commented:
> For the below example, I am getting different molecule build coordinates for every time I run it. I'm not sure if we need/expect the identical build every time? Maybe I missed it, but is there a random seed default or settable? Maybe I'm not running it properly? This may be in the smiles string load itself, not due to this PR.
> 
> **Example (1,3-Dipropylcyclohexane):**
> 
> comp = mb.load('CCCC1CCCC(C1)CCC', smiles=True) # mBuild compound of the monomer unit
> chain = Polymer()
> 
> chain.add_monomer(compound=comp,
> indices=[13, -3],
> separation=.15,
> replace=True)
> 
> chain.build(n=10, sequence='A')
> chain.visualize(show_ports=True).show()

Are you getting the same structure, but different coordinates? It's probably related to #838 which is addressed by the RDkit PR. 

The order of the atoms is slightly different with RDKit though, so if you merge that PR to test it out along with this one, you'll have to double-check what you use for `indices`

bc118 commented:
I changed the comment as I was testing and writing it.  It seems to be related to the build and if trans-cis is stated.  Im not sure its related to the polymer builder itself.

bc118 reviewed (changes requested):
I reviewed the PR and did some of my own testing on it, and I will say I believe this is a great improvement on the polymer builder.  

I make a few comments with the docstrings that should be a little more descriptive.  The only part I found that was not automatically fixed/easy, was picking the right atoms to use to replace with ports.  However, I'm not sure how to resolve this in an automatic fashion that would work for all cases, so I believe it is OK as is.  

The only other main comment was the reproducability of the molecules.  However, this appears to be a bigger issue that is being addressed in another PR (see PR comments), so it does not need to be fixed here before merging. 

Once the docstrings are update it should be OK to approve.

bc118 reviewed (changes requested):
The coverage for the code should be closer to the base code as well before final approval.

justinGilmer commented on mbuild/lib/recipes/polymer.py:
```diff
@@ -1,75 +1,296 @@
+        self._monomers = monomers or []
+        self._end_groups = end_groups or [None, None]
+        if len(self._end_groups) != 2:
+            raise ValueError(
```
I think once you add some unit tests to ensure that these errors are being properly raised when encountered, this should be good to go!

chrisjonesBSU replied:
Oops, I added some tests, but not ones that check for these errors.  I'll be sure to add these as well.

justinGilmer reviewed (changes requested):
I think once you are able to test for some of these errors that codecov has not found covered in the unit tests yet, this should be good to go!

chrisjonesBSU commented:
I added a few unit tests and removed the jupyter notebook.  I'll clean it up some, add more detail about what is going on behind the scenes and make a PR to the tutorial repo.

chrisjonesBSU commented:
I added some relatively long notes to the doc strings. I'm not sure if there are certain standards to follow when adding longer notes to doc strings, so let me know if anything needs to be changed.

CalCraven commented:
Got a chance to play around with this, it's really awesome! Very easy to use. Only one suggestion for the `chain.build` docstring:
>sequence : str, optional, default 'A'
    >A string of characters where each unique character represents one
    >repetition of a monomer. Characters in `sequence` are assigned to
    >monomers in the order they appear in `Polymer.monomers`.

I think it should be a little bit clearer for the ordering that the monomers will be placed in. As far as I can tell with the current implementation, the first monomer added is the `A` monomer, the second would be `B`, etc. That might seem intuitive, but who knows how people will interpret that. I would just add something to the effect of "Characters in `sequence` are assigned to monomers in the order they appear in `Polymer.monomers` i.e. `A` being the first monomer added and B being the second."

jennyfothergill commented on mbuild/lib/recipes/polymer.py:
```diff
@@ -45,40 +148,241 @@ def __init__(self, monomers, n, sequence="A", port_labels=("up", "down")):
         for n_added, seq_item in enumerate(it.cycle(sequence)):
             this_part = clone(seq_map[seq_item])
             self.add(this_part, "monomer[$]")
             if last_part is None:
                 first_part = this_part
             else:
```
`first_part` is never used. I think we can simplify this logic to:
```suggestion
            if last_part is not None:
```

jennyfothergill reviewed (approved):
I've played with this PR a lot an am a fan. c: I suggest one minor change.

chrisjonesBSU commented:
FYI, for anyone that still wants/needs to review this, I made a PR in the mbuild-examples repo that goes over the various ways to build polymers.  

https://github.com/mosdef-hub/mbuild-examples/pull/6

justinGilmer reviewed (approved):
Based on the example you added in mbuild-examples and the notebooks and tests here, this looks like it is ready to go!

bc118 reviewed (changes requested):
I approve with my requested changes
